File: LinkedList/CircularLinkedLists/CircularLinkedList.java
public class CircularLinkedList {
    private static class Node {
        int value;
        Node next;
        Node(int value) {
            this.value = value;
            this.next = null;
        }
    }

    private Node head = null;

    public void prepend(int value) {
        Node newNode = new Node(value);
        if (head == null) {
            newNode.next = newNode;
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != head) {
            current = current.next;
        }
        newNode.next = head;
        current.next = newNode;
        head = newNode;
    }

    public void append(int value) {
        Node newNode = new Node(value);
        if (head == null) {
            newNode.next = newNode;
            head = newNode;
            return;
        }
        Node current = head;
        while (current.next != head) {
            current = current.next;
        }
        current.next = newNode;
        newNode.next = head;
    }

    public void insert(int value, int position) {
        if (position == 0) {
            prepend(value);
            return;
        }
        Node newNode = new Node(value);
        Node current = head;
        for (int i = 0; i < position - 1; i++) {
            current = current.next;
            if (current == head) return;
        }
        newNode.next = current.next;
        current.next = newNode;
    }

    public void deleteFirst() {
        if (head == null) return;
        if (head.next == head) {
            head = null;
            return;
        }
        Node current = head;
        while (current.next != head) {
            current = current.next;
        }
        current.next = head.next;
        head = head.next;
    }

    public void deleteLast() {
        if (head == null) return;
        if (head.next == head) {
            head = null;
            return;
        }
        Node current = head;
        Node prev = null;
        while (current.next != head) {
            prev = current;
            current = current.next;
        }
        prev.next = head;
    }

    public void delete(int position) {
        if (head == null) return;
        if (position == 0) {
            deleteFirst();
            return;
        }
        Node current = head;
        Node prev = null;
        for (int i = 0; i < position; i++) {
            prev = current;
            current = current.next;
            if (current == head) return;
        }
        prev.next = current.next;
    }

    public int search(int value) {
        if (head == null) return -1;
        Node current = head;
        int position = 0;
        do {
            if (current.value == value) return position;
            current = current.next;
            position++;
        } while (current != head);
        return -1;
    }

    public void print() {
        if (head == null) {
            System.out.println("Empty List");
            return;
        }
        Node current = head;
        do {
            System.out.print(current.value + " -> ");
            current = current.next;
        } while (current != head);
        System.out.println("(head)");
    }

    public static void main(String[] args) {
        CircularLinkedList cll = new CircularLinkedList();
        cll.append(10);
        cll.append(20);
        cll.prepend(5);
        cll.insert(15, 2);
        cll.print(); // 5 -> 10 -> 15 -> 20 -> (head)
    }
}

File: LinkedList/CircularLinkedLists/CircularLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.head = None

    def prepend(self, value):
        new_node = Node(value)
        if not self.head:
            new_node.next = new_node
            self.head = new_node
            return
        
        current = self.head
        while current.next != self.head:
            current = current.next
        
        new_node.next = self.head
        current.next = new_node
        self.head = new_node

    def append(self, value):
        new_node = Node(value)
        if not self.head:
            new_node.next = new_node
            self.head = new_node
            return
        
        current = self.head
        while current.next != self.head:
            current = current.next
        
        current.next = new_node
        new_node.next = self.head

    def insert(self, value, position):
        if position == 0:
            self.prepend(value)
            return
        
        new_node = Node(value)
        current = self.head
        for _ in range(position - 1):
            current = current.next
            if current == self.head:
                raise IndexError("Position out of bounds")
        
        new_node.next = current.next
        current.next = new_node

    def deleteFirst(self):
        if not self.head:
            return
        if self.head.next == self.head:
            self.head = None
            return
        
        current = self.head
        while current.next != self.head:
            current = current.next
            
        current.next = self.head.next
        self.head = self.head.next

    def deleteLast(self):
        if not self.head:
            return
        if self.head.next == self.head:
            self.head = None
            return
        
        current = self.head
        prev = None
        while current.next != self.head:
            prev = current
            current = current.next
            
        prev.next = self.head

    def delete(self, position):
        if not self.head:
            return
        if position == 0:
            self.deleteFirst()
            return
        
        current = self.head
        prev = None
        for _ in range(position):
            prev = current
            current = current.next
            if current == self.head:
                return
                
        prev.next = current.next

    def search(self, value):
        if not self.head:
            return -1
        
        current = self.head
        position = 0
        while True:
            if current.value == value:
                return position
            current = current.next
            position += 1
            if current == self.head:
                break
        return -1

    def print(self):
        if not self.head:
            print("Empty List")
            return
        
        current = self.head
        elements = []
        while True:
            elements.append(str(current.value))
            current = current.next
            if current == self.head:
                break
        print(" -> ".join(elements) + " -> (head)")

if __name__ == "__main__":
    cll = CircularLinkedList()
    cll.append(10)
    cll.append(20)
    cll.prepend(5)
    cll.insert(15, 2)
    cll.print()  # Output: 5 -> 10 -> 15 -> 20 -> (head)
    cll.delete(2)
    cll.print()  # Output: 5 -> 10 -> 20 -> (head)
